custom_components/smartthinq_sensors/wideq/device.py

```diff
@@ -843,8 +894,11 @@ def init_device_info(self):
 
             model_data = self._model_data
             if "Monitoring" in model_data and "Value" in model_data:
-                # this are old V1 model
-                self._model_info = ModelInfo(model_data)
+                if isinstance(model_data["Monitoring"]["protocol"], list):
+                    self._model_info = ModelInfoV1Oven(model_data)
+                else:
```

Comment:
This is present also in others old "V1" device, so this check will break support for others devices. This `ModelInfoV1Oven` seems another info type used by some devices (like Oven) based on APIv2 snapshot. We should found a way to manage this without breaking existing old models. Also we should understand how to manage V1 oven. Better solution should be extend `ModelInfo` to manage all models with root keys `Monitoring` and `Values` in the right way.

Reply:
Thanks for the feedback. I don't have any other "V1" devices so it is hard for me to generalize the existing `ModelInfo`. What I do know is that the existing `ModelInfo` clearly does not work for the oven. I will give you more information here and then we can have a conversation about how to better update the existing `ModelInfo`.

First, let's look at protocol `["Monitoring"]["protocol"]`. The existing code assumes the protocol is a `dict` (https://github.com/ollo69/ha-smartthinq-sensors/blob/master/custom_components/smartthinq_sensors/wideq/device.py#L483). But in the case of an oven, it is a list of `dict`s:

```
[
   {
      "_comment":"Upper Oven Operation State(4byte)",
      "superSet":"ovenState.upperState",
      "value":"UpperOvenState"
   }, 
  ...
   {
      "_comment":"Center Burner Hour",
      "superSet":"ovenState.cooktop3OperationTimeHour",
      "value":"CenterHour"
   },
   {
      "_comment":"Drawer Power Level",
      "superSet":"ovenState.drawerPowerLevel",
      "value":"DrawerPowetLevel"
   }
]
```

The `superSet` field refers to the path of the value in the `snapshot`. The `value` field refers to the value type.

Second, the existing `enum_name()` method does not work for ovens. The values `["Value"]` are indexed by value types and not by value keys. Yet, we are calling `enum_name()` with the key and not the value type:

```
{
   "UpperOvenState":{
      "type":"Enum",
      "default":"INITIAL",
      "option":{
         "INITIAL":"@OV_STATE_INITIAL_W",
         "PREHEATING":"@OV_STATE_PREHEAT_W",
         "COOKING_IN_PROGRESS":"@OV_STATE_COOK_W",
         "DONE":"@OV_STATE_COOK_COMPLETE_W",
         "COOLING":"@OV_TERM_COOLING_W"
      }
   },
  ...
``` 

Hence, to get the value type from a value key, you need to look at the protocol to find the value type name from the key name. That's why I added the `value_type_for_key()` method. FWIW, `enum_value()`, `bit_name()`, and `reference_name()` should have the same problem, but they are not being used for the oven.

Do you have access to V1 devices?

Reply:
> First, let's look at protocol `["Monitoring"]["protocol"]`. The existing code assumes the protocol is a `dict`

If you check the method `decode_monitor_binary` you will see that this is using a list of dict for `["Monitoring"]["protocol"]`. This is the method used by all V1 devices.
Not really sure that `decode_snapshot` in `ModelInfo` is used at all, I should have added a comment in the past (my fault). I think that I created this for some reason but I don't remember anymore, I will do additional check.

> Second, the existing `enum_name()` method does not work for ovens. The values `["Value"]` are indexed by value types and not by value keys.

I do not really understand here the difference here, looks to me identical to other V1 devices. Anyway we should find the rule to select the proper method at run-time, that is not the struct used by the `protocol` key.

> Do you have access to V1 devices?

 I have a V1 washer, not a V1 oven. We should use issue #98 to ask for information file.

Reply:
The oven is using `decode_snapshot()` because the `platformType` is `thinkq2`. I'll be happy to make this dynamic based on `dict` or `list of dict` inside `decode_snapshot()`.

As for `enum_name()`, I don't understand how this could have possibly worked for other V1 devices unless the `["Value"]` dict looks completely different. Do you have an example for a V1 `["Value"]` dict that is indexed by key instead of value name?

I'm assuming there is no official documentation from LG how this is all supposed to work?

Reply:
I understand now how I can make this all cleaner. Let me rework the pull request.

Also, what do you think about adding some unit tests? The original wideq repo has unit tests. If we added a `fixtures` directory with output from different LG devices, that would help development going forward. Especially, since the LG API since so weird and not documented.

Reply:
I have reworked the pull request. I will add unit tests next if you agree it is a good idea.

Reply:
Now looks good, great job! I still have a couple of minor review that I will add shortly.
For unit test I'm not very expert, I agree to add, but I would like to use a separate branch for this. Let's close this PR and then I will create a new branch to manage this.

Reply:
Ok, I will create a separate new branch for unit tests.